collector/src/main/java/com/bitmovin/analytics/retryBackend/RetryBackend.kt

```diff
@@ -0,0 +1,162 @@
+
+                override fun onFailure(call: Call, e: IOException) {
+                    if (e is SocketTimeoutException || e is ConnectException || e is StreamResetException || e is UnknownHostException) {
+                        addSample(retrySample)
+                        call.cancel()
+                    }
+                }
+
+                override fun onResponse(call: Call, response: Response) {
+                }
+            })
+        } else if (retrySample.adEventData != null) {
+            retrySample.adEventData.retry = retrySample.retry
+            this.next.sendAd(retrySample.adEventData, object : Callback {
+                override fun onFailure(call: Call, e: IOException) {
+
+                    if (e is SocketTimeoutException) {
+                        addSample(retrySample)
+                    }
+                }
+
+                override fun onResponse(call: Call, response: Response) {
+                    processQueuedSamples()
+                }
+            })
+        }
+    }
+
+    @Synchronized
+    fun processQueuedSamples() {
+        try {
+            if (retrySamplesSet.size > 0) {
+                val nextScheduledTime = retrySamplesSet.first().scheduledTime
+
+                if (retryToken != nextScheduledTime) {
+                    if (retryToken != null) {
+                        handler.removeCallbacks(processSampleRunnable, retryToken)
+                    }
+                    retryToken = nextScheduledTime
+                    val delay = maxOf(nextScheduledTime.time - Date().time, 0) // to prevent negative delay
+                    val t = handler.postAtTime(processSampleRunnable, retryToken, SystemClock.uptimeMillis() + delay)
+                    Log.d(TAG, t.toString())
+                }
+            }
+        } catch (e: Exception) {
+            Log.d(TAG, "processQueuedSamples exception ${e.message}")
+        }
+    }
+
+    private val processSampleRunnable = Runnable {
+        try {
+            val retrySample = getSample()
+            Log.d(TAG, "pop sample ${retrySample?.eventData?.sequenceNumber}")
+            if (retrySample != null) {
+                scheduleSample(retrySample)
+            }
+            retryToken = null
+            processQueuedSamples()
+        } catch (e: Exception) {
+            Log.d(TAG, "processSampleRunnable exception ${e.message}")
+        }
+    }
+
+    fun addSample(retrySample: RetrySample) {
+
+        try {
+            lock.lock()
+            retrySample.retry++
+            val backOffTime = minOf(2.toDouble().pow(retrySample.retry).toInt(), 64)
+            // more than 5min in queue
+            if (retrySample.totalTime + backOffTime < MAX_RETRY_TIME) {
+
+                retrySample.scheduledTime = Calendar.getInstance().run {
+                    add(Calendar.SECOND, backOffTime)
+                    time
+                }
+                retrySample.totalTime += backOffTime
```

Comment:
you can just move that before the `if`, as we will discard the sample anyways if it's too old - so you don't have to do the calculation twice (I know it's just a simple `+`, but in case it changes in future)